### DISCUSSION TITLE
Relational API and SQL injection

### DISCUSSION THREAD

**Discussion by esriis** [duckdb/duckdb#8242] [General]
I am curious about relational objects in python and the risks of using formatted strings in the methods `filter` and `project` in terms of SQL injection (as, unless I've missed something, it's not possible to use prepared statements with these methods).

Suppose I have a relation obtained via `connection.table("example_table")`, where the connection is read-only, and the only methods exposed to end users are `filter` and `project`. 

1. If an attacker has full access to the relation, can they use it to access other tables in the database? Or would any damage or breach from a hypothetical SQL injection attack be limited to the relation itself?
2. Does the DuckDB expression parser for the `filter` and `project` methods ensure that only filtered and projected relations can be returned, and without side effects, regardless of the string expression that is passed?

I have tried to look through the source code to get a basic understanding, but it has been slow going as I am fairly new to C++.

**Comment by rzepinskip** [comment]
I believe it is potentially vulnerable - depends how you use Relational API.

For basic string sql injection with `' OR '1'='1` you may use the string value directly in your filter function (and then you are vulnerable) or you may use Expression API to model your condition (and then you should be safe). 

PS That is an opinion based on quick check without looking into code internals.

![image](https://github.com/user-attachments/assets/bb0f696e-3a77-47ef-8e96-c405237bd6c5)
